File: benchmarks/scroll/src/main/scala/bank/Transaction.scala
package bank

import scroll.internal.compartment.impl.Compartment
import scroll.internal.dispatch.DispatchQuery

case class Transaction() extends Compartment {

  roleGraph.reconfigure(true, false)

  def execute(amount: Float): Unit = {
    roleQueries.one[Source]().withDraw(amount)
    roleQueries.one[Target]().deposit(amount)
  }

  case class Source() {

    def withDraw(m: Float): Unit = {
      // println("Source.withDraw")
      val _ = (+this).decrease(m)
    }

  }

  case class Target() {

    def deposit(m: Float): Unit = {
      // println("Target.deposit")
      val _ = (+this).increase(m)
    }

  }

}

File: benchmarks/role4j/src/main/java/bank/Transaction.java
package bank;

import net.role4j.IPlayer;
import net.role4j.IRole;

public class Transaction implements IPlayer {

  private Account source;
  private Account target;

  public void setAccounts(Account source, Account target) {
    try {
      source.bind(Source.class);
      target.bind(Target.class);
    } catch (Throwable e) {
      //
    }
    this.source = source;
    this.target = target;
  }

  public void execute(float amount) {
    source.decrease(amount);
    target.increase(amount);
  }
}
